src/Proto3/Wire/Encode.hs

```diff
@@ -72,25 +73,81 @@ module Proto3.Wire.Encode
-    , BB.toLazyByteString
+    , Builder
+    , builderLength
+    , toLazyByteString
     ) where
 
-import           Data.Bits                   ( (.&.), (.|.), shiftL, shiftR, xor )
-import qualified Data.ByteString             as B
-import qualified Data.ByteString.Builder     as BB
-import qualified Data.ByteString.Lazy        as BL
-import           Data.Int                    ( Int32, Int64 )
-import           Data.Monoid                 ( (<>) )
-import qualified Data.Text.Lazy              as Text.Lazy
-import qualified Data.Text.Lazy.Encoding     as Text.Lazy.Encoding
-import           Data.Word                   ( Word32, Word64, Word8 )
+import           Data.Bits                     ( (.&.), (.|.), shiftL, shiftR, xor )
+import qualified Data.ByteString               as B
+import qualified Data.ByteString.Builder       as BB
+import qualified Data.ByteString.Builder.Extra as BB
+import qualified Data.ByteString.Lazy          as BL
+import           Data.Char                     ( ord )
+import           Data.Int                      ( Int32, Int64 )
+import           Data.Monoid                   ( Sum(..), (<>) )
+import qualified Data.Text.Encoding            as Text.Encoding
+import qualified Data.Text.Lazy                as Text.Lazy
+import qualified Data.Text.Lazy.Encoding       as Text.Lazy.Encoding
+import           Data.Word                     ( Word32, Word64, Word8 )
 import           Proto3.Wire.Types
 
-base128Varint :: Word64 -> BB.Builder
+-- | Like 'BB.Builder', but memoizes the resulting length so
+-- that we can efficiently encode nested embedded messages.
+newtype Builder = Builder { unBuilder :: (Sum Word, BB.Builder) }
+  deriving Monoid
+
+builderLength :: Builder -> Word
+builderLength = getSum . fst . unBuilder
+
+toLazyByteString :: Builder -> BL.ByteString
+toLazyByteString (Builder (Sum len, bb)) =
+    BB.toLazyByteStringWith strat BL.empty bb
+  where
+    -- If the supplied length is accurate then we will perform just
+    -- one allocation (unless that length exceeds 'maxFirstChunk').
+    -- An inaccurate length would indicate a bug in
+    -- one of the primitives that produces a 'Builder'.
+    maxFirstChunk = 134217728  -- 128MiB
+    firstChunk = fromIntegral (min maxFirstChunk len)
+    strat = BB.safeStrategy firstChunk BB.defaultChunkSize
+{-# NOINLINE toLazyByteString #-}
+
+word8 :: Word8 -> Builder
+word8 w = Builder (Sum 1, BB.word8 w)
+
+word32LE :: Word32 -> Builder
+word32LE w = Builder (Sum 4, BB.word32LE w)
+
+int32LE :: Int32 -> Builder
+int32LE w = Builder (Sum 4, BB.int32LE w)
+
+floatLE :: Float -> Builder
+floatLE f = Builder (Sum 4, BB.floatLE f)
+
+word64LE :: Word64 -> Builder
+word64LE w = Builder (Sum 8, BB.word64LE w)
+
+int64LE :: Int64 -> Builder
+int64LE w = Builder (Sum 8, BB.int64LE w)
+
+doubleLE :: Double -> Builder
+doubleLE f = Builder (Sum 8, BB.doubleLE f)
+
+stringUtf8 :: String -> Builder
+stringUtf8 s = Builder (Sum (len 0 s), BB.stringUtf8 s)
```

Comment:
Well, it shouldn't build up because you can't nest these calls, so it will only ever be 2 passes, and the first one shouldn't allocate significant memory.

What would the warning say?